src/openlifu/vf/virtual_fit.py

```diff
@@ -0,0 +1,161 @@
+        """
+        Analyse the transducer position relative to a specific target.
+        """
+        #TODO: Compute if target is within steering limits
+        #TODO: In the future, we should implement the ray-tracing analysis given a full segmentation
+
+        # pos_analysis = 1.0
+        # target_tr_space = target2trspace(pos, target)
+        # for target_constraint in self.steering_limits:
+        #     pos = target_tr_space.get_position(
+        #         dim=target_constraint.dim,
+        #         units=target_constraint.units
+        #     )
+        #     try:
+        #         target_constraint.check_bounds(pos)
+        #     except ValueError:
+        #         pos_analysis = 0.0
+        #
+        # return pos_analysis
+
+        pass
+
+    def run(
+            self,
+            target: Point,
+            pitch_range: Optional[Tuple[int, int]] = None,
+            pitch_step: Optional[int] = None,
+            yaw_range: Optional[Tuple[int, int]] = None,
+            yaw_step: Optional[int] = None,
+            steering_limits: Optional[Tuple[TargetConstraints]] = None,
+            blocked_elems_threshold: Optional[float] = None
+        ) -> np.ndarray:
+        """
+        VirtualFit main process.
+
+        Finds the optimal transducer transform (position and orientation)
+        given an input MRI volume and the associated target.
+        """
+        if pitch_range is None:
+            pitch_range = self.pitch_range
+        if pitch_step is None:
+            pitch_step = self.pitch_step
+        if yaw_range is None:
+            yaw_range = self.yaw_range
+        if yaw_step is None:
+            yaw_step = self.yaw_step
+        if steering_limits is None:
+            steering_limits = self.steering_limits
+        if blocked_elems_threshold is None:
+            blocked_elems_threshold = self.blocked_elems_threshold
+
+        self.logger.info("Running VirtualFit main process.")
+        self.logger.info("VirtualFit: Searching optimal position...")
+        # 2. get search grid
+        search_grid = self.get_search_grid(yaw_range, yaw_step, pitch_range, pitch_step)
+        for i in range(search_grid[0].shape[0]):
+            for j in range(search_grid[0].shape[1]):
+                yaw, pitch = (search_grid[0][i, j], search_grid[1][i, j])
+                self.logger.info(f"VirtualFit: Analysing {(yaw, pitch)}...")
+                # 3. define transducer transform (plane fitting) on the surface (skin) given spherical coordinate (yaw, pitch)
+                # self.fit_to_surface(sph_coords: Tuple[float, float], skin_surface: np.ndarray)
+                # 4. analyse current transform
+                # self.analyse_position(pos: np.ndarray, transducer: Transducer, target: Point):
+                optimal_transform = np.zeros((4, 4))
+        self.logger.info("VirtualFit: Found optimal position!")
+
+        return optimal_transform
```

Comment:
I'd assume Volume units, typically mm. In fact, the only significant use of changing units is for setting up the k-wave sims, which have to be m. Otherwise, I'd have just forced it all into mm.

Reply:
Also the transform unit (and tr matrix) is defined in`array_transform` from `subject/sessions/session.json`. So we can convert it later. Just updated the `optimal_transform` typing for clarity. https://github.com/OpenwaterHealth/OpenLIFU-python/pull/162/commits/5594c67057ff754860b698b76149bdcb0bbfa087